{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9561e0b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading datasets from: /mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/\n",
      "\n",
      "--- Loading ADNI ---\n",
      "File: /mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/adni_standardized_preprocessed.h5ad\n",
      "Successfully loaded ADNI with shape (650, 17991)\n",
      "------------------\n",
      "--- Loading ENCODE ---\n",
      "File: /mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/encode_standardized_preprocessed.h5ad\n",
      "Successfully loaded ENCODE with shape (7, 65586)\n",
      "--------------------\n",
      "--- Loading GTEX ---\n",
      "File: /mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/gtex_standardized_preprocessed.h5ad\n",
      "Successfully loaded GTEX with shape (19616, 58988)\n",
      "------------------\n",
      "--- Loading MAGE ---\n",
      "File: /mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/mage_standardized_preprocessed.h5ad\n",
      "Successfully loaded MAGE with shape (731, 19428)\n",
      "------------------\n",
      "\n",
      "\n",
      "========================= Inspecting: ADNI =========================\n",
      "\n",
      "--- AnnData Object Info (ADNI) ---\n",
      "AnnData object with n_obs × n_vars = 650 × 17991\n",
      "    obs: 'sample_id', 'subject_id', 'dataset', 'data_type', 'expression_unit', 'tissue', 'platform', 'processing', 'sex', 'age', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence'\n",
      "    var: 'gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence'\n",
      "    uns: 'dataset_info', 'gencode_version', 'gene_mapping_stats', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_validation', 'ontology_mappings', 'original_gencode_version', 'original_reference_genome', 'processing_date', 'reference_genome', 'subject_demographics'\n",
      "------------------------------\n",
      "\n",
      "--- Observations (.obs) Info (ADNI) ---\n",
      ".obs shape: (650, 19)\n",
      ".obs columns: ['sample_id', 'subject_id', 'dataset', 'data_type', 'expression_unit', 'tissue', 'platform', 'processing', 'sex', 'age', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence']\n",
      "\n",
      ".obs.head():\n",
      "                                                                         sample_id  subject_id dataset   data_type       expression_unit tissue                            platform   processing      sex age species species_ontology tissue_original tissue_ontology data_type_original assay_ontology age_original developmental_stage_ontology tissue_ontology_confidence\n",
      "_sample_id                                                                                                                                                                                                                                                                                                                                                                   \n",
      "002_S_0413_002_S_0413_gencode_v24_pruned  002_S_0413_002_S_0413_gencode_v24_pruned  002_S_0413    ADNI  Microarray  Normalized intensity  blood  Affymetrix Human Genome U219 Array  gencode_v24  unknown       human   NCBITaxon:9606           blood  UBERON:0000178         Microarray                                                                              medium\n",
      "002_S_0729_002_S_0729_gencode_v24_pruned  002_S_0729_002_S_0729_gencode_v24_pruned  002_S_0729    ADNI  Microarray  Normalized intensity  blood  Affymetrix Human Genome U219 Array  gencode_v24  unknown       human   NCBITaxon:9606           blood  UBERON:0000178         Microarray                                                                              medium\n",
      "002_S_1155_002_S_1155_gencode_v24_pruned  002_S_1155_002_S_1155_gencode_v24_pruned  002_S_1155    ADNI  Microarray  Normalized intensity  blood  Affymetrix Human Genome U219 Array  gencode_v24  unknown       human   NCBITaxon:9606           blood  UBERON:0000178         Microarray                                                                              medium\n",
      "002_S_1261_002_S_1261_gencode_v24_pruned  002_S_1261_002_S_1261_gencode_v24_pruned  002_S_1261    ADNI  Microarray  Normalized intensity  blood  Affymetrix Human Genome U219 Array  gencode_v24  unknown       human   NCBITaxon:9606           blood  UBERON:0000178         Microarray                                                                              medium\n",
      "002_S_1268_002_S_1268_gencode_v24_pruned  002_S_1268_002_S_1268_gencode_v24_pruned  002_S_1268    ADNI  Microarray  Normalized intensity  blood  Affymetrix Human Genome U219 Array  gencode_v24  unknown       human   NCBITaxon:9606           blood  UBERON:0000178         Microarray                                                                              medium\n",
      "\n",
      "Unique values in key .obs columns:\n",
      "  'dataset': ['ADNI'] (1 total unique)\n",
      "  'tissue': ['blood'] (1 total unique)\n",
      "  'sex': ['unknown'] (1 total unique)\n",
      "  'data_type': ['Microarray'] (1 total unique)\n",
      "  'expression_unit': ['Normalized intensity'] (1 total unique)\n",
      "  'tissue_ontology': ['UBERON:0000178'] (1 total unique)\n",
      "\n",
      "Missing values count in key .obs columns:\n",
      "{'age': np.int64(650)}\n",
      "------------------------------\n",
      "\n",
      "--- Variables (.var) Info (ADNI) ---\n",
      ".var shape: (17991, 9)\n",
      ".var columns: ['gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence']\n",
      ".var index name: None\n",
      "First 5 var_names (index): ['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000938']\n",
      "\n",
      ".var.head():\n",
      "                         gene_id        original_ids gene_name       gene_type chromosome     mapping_source original_gene_id       ensembl_id mapping_confidence\n",
      "ENSG00000000003  ENSG00000000003  ENSG00000000003.14    TSPAN6  protein_coding       chrX  reference_mapping  ENSG00000000003  ENSG00000000003               high\n",
      "ENSG00000000005  ENSG00000000005   ENSG00000000005.5      TNMD  protein_coding       chrX  reference_mapping  ENSG00000000005  ENSG00000000005               high\n",
      "ENSG00000000419  ENSG00000000419  ENSG00000000419.12      DPM1  protein_coding      chr20  reference_mapping  ENSG00000000419  ENSG00000000419               high\n",
      "ENSG00000000457  ENSG00000000457  ENSG00000000457.13     SCYL3  protein_coding       chr1  reference_mapping  ENSG00000000457  ENSG00000000457               high\n",
      "ENSG00000000938  ENSG00000000938  ENSG00000000938.12       FGR  protein_coding       chr1  reference_mapping  ENSG00000000938  ENSG00000000938               high\n",
      "\n",
      "Unique values in key .var columns:\n",
      "  'gene_type': ['protein_coding', 'unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'antisense', 'processed_pseudogene', 'unprocessed_pseudogene', 'lincRNA', 'transcribed_processed_pseudogene', 'transcribed_unitary_pseudogene', 'polymorphic_pseudogene']... (23 total unique)\n",
      "  'chromosome': ['chrX', 'chr20', 'chr1', 'chr6', 'chr3', 'chr7', 'chr12', 'chr11', 'chr4', 'chr17']... (24 total unique)\n",
      "  'mapping_source': ['reference_mapping'] (1 total unique)\n",
      "  'mapping_confidence': ['high'] (1 total unique)\n",
      "------------------------------\n",
      "\n",
      "--- Unstructured Annotation (.uns) Info (ADNI) ---\n",
      ".uns keys: ['dataset_info', 'gencode_version', 'gene_mapping_stats', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_validation', 'ontology_mappings', 'original_gencode_version', 'original_reference_genome', 'processing_date', 'reference_genome', 'subject_demographics']\n",
      "\n",
      ".uns['dataset_info']: \n",
      "{\n",
      "  \"assay_ontology\": \"EFO:0002695\",\n",
      "  \"data_type\": \"microarray\",\n",
      "  \"description\": \"Alzheimer's Disease Neuroimaging Initiative microarray data with demographic information\",\n",
      "  \"expression_unit\": \"quantile normalized microarray counts\",\n",
      "  \"gencode_version\": \"24\",\n",
      "  \"platform\": \"Affymetrix Human Genome U219 Array\",\n",
      "  \"reference\": \"https://pmc.ncbi.nlm.nih.gov/articles/PMC7541709/\",\n",
      "  \"reference_genome\": \"hg38\",\n",
      "  \"source\": \"ADNI\",\n",
      "  \"version\": \"April 2025\"\n",
      "}\n",
      "\n",
      ".uns['gene_mapping_stats']: \n",
      "{\n",
      "  \"mapped_genes\": \"17991\",\n",
      "  \"mapping_percentage\": 100.0,\n",
      "  \"mapping_source\": \"reference_mapping\",\n",
      "  \"total_genes\": \"17991\"\n",
      "}\n",
      "\n",
      ".uns['metadata_validation']: \n",
      "{\n",
      "  \"dataset\": \"ADNI\",\n",
      "  \"missing_fields\": \"[]\",\n",
      "  \"missing_values\": {},\n",
      "  \"sample_count\": \"650\",\n",
      "  \"unmapped_ages\": \"[]\",\n",
      "  \"unmapped_assays\": \"['Microarray']\",\n",
      "  \"unmapped_tissues\": \"[]\",\n",
      "  \"validation_status\": \"warning\"\n",
      "}\n",
      "\n",
      ".uns['harmonized_gencode_version']: \n",
      "24\n",
      "\n",
      ".uns['harmonized_reference_genome']: \n",
      "hg38\n",
      "------------------------------\n",
      "\n",
      "\n",
      "========================= Inspecting: ENCODE =========================\n",
      "\n",
      "--- AnnData Object Info (ENCODE) ---\n",
      "AnnData object with n_obs × n_vars = 7 × 65586\n",
      "    obs: 'original_sample_id', 'cell_line', 'description', 'tissue', 'disease', 'cell_type', 'sex', 'organism', 'age', 'ethnicity', 'subject_id', 'geo_id', 'nucleic_acid_type', 'depletion', 'size_range', 'strand_specificity', 'cell_type_info', 'extraction_method', 'data_type', 'expression_unit', 'assay', 'dataset', 'sample_id', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence'\n",
      "    var: 'original_id_from_input', 'original_id_from_source', 'ensembl_id', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'mapping_confidence', 'gene_id'\n",
      "    uns: 'assay_ontology', 'cell_type_info', 'data_type', 'dataset_info', 'gencode_version', 'gene_mapping_stats', 'genome_info', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'ontology_mappings', 'platform', 'processing_date', 'reference_genome', 'rna_seq_protocol'\n",
      "------------------------------\n",
      "\n",
      "--- Observations (.obs) Info (ENCODE) ---\n",
      ".obs shape: (7, 32)\n",
      ".obs columns: ['original_sample_id', 'cell_line', 'description', 'tissue', 'disease', 'cell_type', 'sex', 'organism', 'age', 'ethnicity', 'subject_id', 'geo_id', 'nucleic_acid_type', 'depletion', 'size_range', 'strand_specificity', 'cell_type_info', 'extraction_method', 'data_type', 'expression_unit', 'assay', 'dataset', 'sample_id', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence']\n",
      "\n",
      ".obs.head():\n",
      "            original_sample_id cell_line                                     description       tissue                          disease    cell_type     sex organism age ethnicity   subject_id        geo_id    nucleic_acid_type depletion size_range         strand_specificity                                     cell_type_info extraction_method data_type expression_unit                    assay dataset sample_id species species_ontology tissue_original tissue_ontology data_type_original assay_ontology age_original developmental_stage_ontology tissue_ontology_confidence\n",
      "_sample_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
      "ENCFF244DNJ        ENCFF244DNJ      A549                  Human lung carcinoma cell line         lung                   adenocarcinoma   epithelial    male    human  58  European  ENCDO451RUA  SAMN05733878  polyadenylated mRNA      rRNA       >200  Strand-specific (reverse)  {'anatomical_entity_id': 'UBERON:0002048', 'ce...        polyA_plus   RNA-seq             TPM                           ENCODE             human   NCBITaxon:9606            lung  UBERON:0002048            RNA-seq    EFO:0009922           58               HsapDv:0000087                     medium\n",
      "ENCFF685WJV        ENCFF685WJV     Caki2     Human kidney clear cell carcinoma cell line       kidney  clear cell renal cell carcinoma   epithelial    male    human  69  European  ENCDO869AAI  SAMN04284635                  RNA      rRNA       >200  Strand-specific (reverse)  {'anatomical_entity_id': 'UBERON:0002113', 'ce...                     RNA-seq             TPM  RNA-seq (total RNA-seq)  ENCODE             human   NCBITaxon:9606          kidney  UBERON:0002113            RNA-seq    EFO:0009922           69               HsapDv:0000087                     medium\n",
      "ENCFF640FPG        ENCFF640FPG   GM23248                 Human skin fibroblast cell line         skin                           normal   Fibroblast    male    human  53  European  ENCDO467QPX  SAMN04284514                  RNA      rRNA       >200  Strand-specific (reverse)  {'anatomical_entity_id': 'UBERON:0002097', 'ce...                     RNA-seq             TPM  RNA-seq (total RNA-seq)  ENCODE             human   NCBITaxon:9606            skin  UBERON:0002097            RNA-seq    EFO:0009922           53               HsapDv:0000087                     medium\n",
      "ENCFF863QWG        ENCFF863QWG     HepG2  Human liver hepatocellular carcinoma cell line        liver         hepatocellular carcinoma   epithelial    male    human  15  European  ENCDO886MPB  SAMN04284581                  RNA      rRNA       >200  Strand-specific (reverse)  {'anatomical_entity_id': 'UBERON:0002107', 'ce...                     RNA-seq             TPM  RNA-seq (total RNA-seq)  ENCODE             human   NCBITaxon:9606           liver  UBERON:0002107            RNA-seq    EFO:0009922           15               HsapDv:0000086                     medium\n",
      "ENCFF171FQU        ENCFF171FQU      K562    Human chronic myelogenous leukemia cell line  bone marrow     chronic myelogenous leukemia  lymphoblast  female    human  53   unknown  ENCDO000AAL  SAMN04284550                  RNA      rRNA       >200  Strand-specific (reverse)  {'anatomical_entity_id': 'UBERON:0002371', 'ce...                     RNA-seq             TPM  RNA-seq (total RNA-seq)  ENCODE             human   NCBITaxon:9606     bone marrow  UBERON:0002371            RNA-seq    EFO:0009922           53               HsapDv:0000087                     medium\n",
      "\n",
      "Unique values in key .obs columns:\n",
      "  'dataset': ['ENCODE'] (1 total unique)\n",
      "  'tissue': ['lung', 'kidney', 'skin', 'liver', 'bone marrow', 'pancreas'] (6 total unique)\n",
      "  'sex': ['male', 'female'] (2 total unique)\n",
      "  'data_type': ['RNA-seq'] (1 total unique)\n",
      "  'expression_unit': ['TPM'] (1 total unique)\n",
      "  'tissue_ontology': ['UBERON:0002048', 'UBERON:0002113', 'UBERON:0002097', 'UBERON:0002107', 'UBERON:0002371', 'UBERON:0001264'] (6 total unique)\n",
      "\n",
      "Missing values count in key .obs columns:\n",
      "{'age': np.int64(1)}\n",
      "------------------------------\n",
      "\n",
      "--- Variables (.var) Info (ENCODE) ---\n",
      ".var shape: (65586, 9)\n",
      ".var columns: ['original_id_from_input', 'original_id_from_source', 'ensembl_id', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'mapping_confidence', 'gene_id']\n",
      ".var index name: feature_id\n",
      "First 5 var_names (index): ['ENSG00000166619', 'ENTREZ:12954', 'ENTREZ:12956', 'ENTREZ:12958', 'ENTREZ:12960']\n",
      "\n",
      ".var.head():\n",
      "                original_id_from_input original_id_from_source       ensembl_id gene_name       gene_type chromosome                 mapping_source mapping_confidence          gene_id\n",
      "feature_id                                                                                                                                                                             \n",
      "ENSG00000166619                  10904                   10904  ENSG00000166619     BLCAP  protein_coding      chr20  encode_mapping -> ref_numeric               high  ENSG00000166619\n",
      "ENTREZ:12954                     12954                   12954     ENTREZ:12954                   unknown                            encode_mapping                low     ENTREZ:12954\n",
      "ENTREZ:12956                     12956                   12956     ENTREZ:12956                   unknown                            encode_mapping                low     ENTREZ:12956\n",
      "ENTREZ:12958                     12958                   12958     ENTREZ:12958                   unknown                            encode_mapping                low     ENTREZ:12958\n",
      "ENTREZ:12960                     12960                   12960     ENTREZ:12960                   unknown                            encode_mapping                low     ENTREZ:12960\n",
      "\n",
      "Unique values in key .var columns:\n",
      "  'gene_type': ['protein_coding', 'unknown', 'processed_transcript', 'processed_pseudogene', 'unprocessed_pseudogene', 'snRNA', 'IG_V_pseudogene', 'IG_D_gene', 'TR_J_gene', 'TR_D_gene']... (46 total unique)\n",
      "  'chromosome': ['chr20', '', 'chr1', 'chr3', 'chr17', 'chr5', 'chr22', 'chr7', 'chr16', 'chr11']... (27 total unique)\n",
      "  'mapping_source': ['encode_mapping -> ref_numeric', 'encode_mapping', 'unmapped', 'spike_in'] (4 total unique)\n",
      "  'mapping_confidence': ['high', 'low', 'none'] (3 total unique)\n",
      "------------------------------\n",
      "\n",
      "--- Unstructured Annotation (.uns) Info (ENCODE) ---\n",
      ".uns keys: ['assay_ontology', 'cell_type_info', 'data_type', 'dataset_info', 'gencode_version', 'gene_mapping_stats', 'genome_info', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'ontology_mappings', 'platform', 'processing_date', 'reference_genome', 'rna_seq_protocol']\n",
      "\n",
      ".uns['dataset_info']: \n",
      "{\n",
      "  \"cell_lines\": {\n",
      "    \"A549\": {\n",
      "      \"age\": \"58\",\n",
      "      \"cell_type\": \"epithelial\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002048\",\n",
      "        \"cell_type_id\": \"CL:0000066\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human lung carcinoma cell line\",\n",
      "      \"disease\": \"adenocarcinoma\",\n",
      "      \"ethnicity\": \"European\",\n",
      "      \"geo_id\": \"SAMN05733878\",\n",
      "      \"nucleic_acid_type\": \"polyadenylated mRNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO451RUA\",\n",
      "      \"tissue\": \"lung\"\n",
      "    },\n",
      "    \"Caki2\": {\n",
      "      \"age\": \"69\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"epithelial\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002113\",\n",
      "        \"cell_type_id\": \"CL:0000066\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human kidney clear cell carcinoma cell line\",\n",
      "      \"disease\": \"clear cell renal cell carcinoma\",\n",
      "      \"ethnicity\": \"European\",\n",
      "      \"geo_id\": \"SAMN04284635\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO869AAI\",\n",
      "      \"tissue\": \"kidney\"\n",
      "    },\n",
      "    \"GM23248\": {\n",
      "      \"age\": \"53\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"Fibroblast\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002097\",\n",
      "        \"cell_type_id\": \"CL:0000057\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human skin fibroblast cell line\",\n",
      "      \"disease\": \"normal\",\n",
      "      \"ethnicity\": \"European\",\n",
      "      \"geo_id\": \"SAMN04284514\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO467QPX\",\n",
      "      \"tissue\": \"skin\"\n",
      "    },\n",
      "    \"HepG2\": {\n",
      "      \"age\": \"15\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"epithelial\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002107\",\n",
      "        \"cell_type_id\": \"CL:0000066\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human liver hepatocellular carcinoma cell line\",\n",
      "      \"disease\": \"hepatocellular carcinoma\",\n",
      "      \"ethnicity\": \"European\",\n",
      "      \"geo_id\": \"SAMN04284581\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO886MPB\",\n",
      "      \"tissue\": \"liver\"\n",
      "    },\n",
      "    \"K562\": {\n",
      "      \"age\": \"53\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"lymphoblast\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002371\",\n",
      "        \"cell_type_id\": \"CL:0000945\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human chronic myelogenous leukemia cell line\",\n",
      "      \"disease\": \"chronic myelogenous leukemia\",\n",
      "      \"ethnicity\": \"unknown\",\n",
      "      \"geo_id\": \"SAMN04284550\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"female\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO000AAL\",\n",
      "      \"tissue\": \"bone marrow\"\n",
      "    },\n",
      "    \"NCI-H460\": {\n",
      "      \"age\": \"\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"epithelial\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0002048\",\n",
      "        \"cell_type_id\": \"CL:0000066\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human lung large cell carcinoma cell line\",\n",
      "      \"disease\": \"large cell lung carcinoma\",\n",
      "      \"ethnicity\": \"unknown\",\n",
      "      \"geo_id\": \"SAMN07791353\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO105AAA\",\n",
      "      \"tissue\": \"lung\"\n",
      "    },\n",
      "    \"Panc1\": {\n",
      "      \"age\": \"56\",\n",
      "      \"assay\": \"RNA-seq (total RNA-seq)\",\n",
      "      \"cell_type\": \"ductal cell\",\n",
      "      \"cell_type_info\": {\n",
      "        \"anatomical_entity_id\": \"UBERON:0001264\",\n",
      "        \"cell_type_id\": \"CL:0000068\"\n",
      "      },\n",
      "      \"depletion\": \"rRNA\",\n",
      "      \"description\": \"Human pancreatic ductal carcinoma cell line\",\n",
      "      \"disease\": \"pancreatic carcinoma\",\n",
      "      \"ethnicity\": \"European\",\n",
      "      \"geo_id\": \"SAMN05733879\",\n",
      "      \"nucleic_acid_type\": \"RNA\",\n",
      "      \"organism\": \"human\",\n",
      "      \"sex\": \"male\",\n",
      "      \"size_range\": \">200\",\n",
      "      \"strand_specificity\": \"Strand-specific (reverse)\",\n",
      "      \"subject_id\": \"ENCDO845WKR\",\n",
      "      \"tissue\": \"pancreas\"\n",
      "    }\n",
      "  },\n",
      "  \"data_repository\": \"[{\\\"name\\\": \\\"ENCODE Project\\\", \\\"url\\\": \\\"https://www.encodeproject.org/\\\"}]\",\n",
      "  \"project\": \"ENCODE\"\n",
      "}\n",
      "\n",
      ".uns['gene_mapping_stats']: \n",
      "{\n",
      "  \"mapped_genes\": \"43086\",\n",
      "  \"mapping_percentage\": 65.69389808800659,\n",
      "  \"mapping_sources\": {\n",
      "    \"encode_mapping\": \"42707\",\n",
      "    \"encode_mapping -> ref_numeric\": \"282\",\n",
      "    \"spike_in\": \"97\",\n",
      "    \"unmapped\": \"22500\"\n",
      "  },\n",
      "  \"total_genes\": \"65586\",\n",
      "  \"unmapped_count\": \"22500\"\n",
      "}\n",
      "\n",
      ".uns['metadata_validation']: \n",
      "{\n",
      "  \"dataset\": \"ENCODE\",\n",
      "  \"missing_fields\": \"[]\",\n",
      "  \"missing_values\": {\n",
      "    \"assay\": \"1\",\n",
      "    \"extraction_method\": \"6\"\n",
      "  },\n",
      "  \"sample_count\": \"7\",\n",
      "  \"unmapped_ages\": \"[]\",\n",
      "  \"unmapped_assays\": \"[]\",\n",
      "  \"unmapped_tissues\": \"[]\",\n",
      "  \"validation_status\": \"warning\"\n",
      "}\n",
      "\n",
      ".uns['harmonized_gencode_version']: \n",
      "24\n",
      "\n",
      ".uns['harmonized_reference_genome']: \n",
      "hg38\n",
      "------------------------------\n",
      "\n",
      "\n",
      "========================= Inspecting: GTEX =========================\n",
      "\n",
      "--- AnnData Object Info (GTEX) ---\n",
      "AnnData object with n_obs × n_vars = 19616 × 58988\n",
      "    obs: 'SMATSSCR', 'SMCENTER', 'SMPTHNTS', 'rna_integrity_number', 'broad_tissue', 'tissue', 'SMUBRID', 'ischemic_time', 'SMTSPAX', 'array_batch', 'SMNABTCHT', 'SMNABTCHD', 'batch', 'SMGEBTCHD', 'SMGEBTCHT', 'ANALYTE_TYPE', 'SMAFRZE', 'SMGTC', 'SMRDTTL', 'SMALTTL', 'SMALTALG', 'SMSUPALG', 'SMRDLGTH', 'SMVQCFL', 'SMLMAPQ', 'SMUMPRD', 'SMUNPDRD', 'SMMPPD', 'SMMAPRT', 'SMMPPDUN', 'SMUNMPRT', 'SMMPDP', 'SMDPMPRT', 'SMMPPDXG', 'SMMPDPXG', 'SMDPRTXG', 'SMCHMRD', 'SMCHMRT', 'SMMPPDPR', 'SMMPHQRD', 'SMMPHQRT', 'SMMPLQRD', 'SMSPLTRT', 'SME1MPRD', 'SME2MPRD', 'SME1MPRT', 'SME2MPRT', 'SME1MMB', 'SME2MMB', 'SME1TTLB', 'SME2TTLB', 'SME1MMRT', 'SME2MMRT', 'SMTTLMM', 'SMTTLB', 'SMBSMMRT', 'SMESTLBS', 'SMEXNCRD', 'SMEXNCRT', 'SMEXPEFF', 'SMNTRNRD', 'SMNTRNRT', 'SMNTRARD', 'SMNTRART', 'SMNTERRD', 'SMNTERRT', 'SMAMBRD', 'SMAMBRT', 'SMNTEXC', 'SMDSCRT', 'SMEXNCRTHQ', 'SMNTRNRTHQ', 'SMNTRARTHQ', 'SMNTERRTHQ', 'SMAMBRTHQ', 'SME1SNSE', 'SME2SNSE', 'SME1ANTI', 'SME2ANTI', 'SME1PCTS', 'SME2PCTS', 'SMGNSDTC', 'SMRRNARD', 'SMRRNART', 'SMMFLGTH', 'SMSFLGTH', 'SMMDFLGTH', 'SMSMFLGTH', 'SMFGCMN', 'SMFGCSD', 'SMFGCSK', 'SMFGCKT', 'SM3PBMN', 'SM3PBSD', 'SM3PBMD', 'SM3PB25P', 'SM3PB75P', 'SM3PBSDM', 'SM3PBGN', 'SMMDMNCV', 'SMMDCVSD', 'SMMDCVCV', 'SMEXCVMD', 'SMEXCVMAD', 'SMMNCV', 'SMUVCRD', 'SMUVCRT', 'SMSHRTRD', 'SMSHRTRT', 'SMSMRDHQ', 'SMSMRTHQ', 'SMPRERDHQ', 'SMPRERTHQ', 'SMSMGNDT', 'SMPREGNDT', 'SMRDLNMN', 'SMRDLNMD', 'SMRDLNSD', 'subject_id', 'sex', 'age', 'DTHHRDY', 'dataset', 'original_sample_id', 'data_type', 'expression_unit', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'original_obs_sample_id', 'tissue_ontology_confidence'\n",
      "    var: 'gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence'\n",
      "    uns: 'dataset_info', 'dataset_version', 'gencode_version', 'gene_mapping_stats', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'original_gencode_version', 'original_reference_genome', 'processing_date', 'reference_genome', 'rna_seq_protocol'\n",
      "------------------------------\n",
      "\n",
      "--- Observations (.obs) Info (GTEX) ---\n",
      ".obs shape: (19616, 136)\n",
      ".obs columns: ['SMATSSCR', 'SMCENTER', 'SMPTHNTS', 'rna_integrity_number', 'broad_tissue', 'tissue', 'SMUBRID', 'ischemic_time', 'SMTSPAX', 'array_batch', 'SMNABTCHT', 'SMNABTCHD', 'batch', 'SMGEBTCHD', 'SMGEBTCHT', 'ANALYTE_TYPE', 'SMAFRZE', 'SMGTC', 'SMRDTTL', 'SMALTTL', 'SMALTALG', 'SMSUPALG', 'SMRDLGTH', 'SMVQCFL', 'SMLMAPQ', 'SMUMPRD', 'SMUNPDRD', 'SMMPPD', 'SMMAPRT', 'SMMPPDUN', 'SMUNMPRT', 'SMMPDP', 'SMDPMPRT', 'SMMPPDXG', 'SMMPDPXG', 'SMDPRTXG', 'SMCHMRD', 'SMCHMRT', 'SMMPPDPR', 'SMMPHQRD', 'SMMPHQRT', 'SMMPLQRD', 'SMSPLTRT', 'SME1MPRD', 'SME2MPRD', 'SME1MPRT', 'SME2MPRT', 'SME1MMB', 'SME2MMB', 'SME1TTLB', 'SME2TTLB', 'SME1MMRT', 'SME2MMRT', 'SMTTLMM', 'SMTTLB', 'SMBSMMRT', 'SMESTLBS', 'SMEXNCRD', 'SMEXNCRT', 'SMEXPEFF', 'SMNTRNRD', 'SMNTRNRT', 'SMNTRARD', 'SMNTRART', 'SMNTERRD', 'SMNTERRT', 'SMAMBRD', 'SMAMBRT', 'SMNTEXC', 'SMDSCRT', 'SMEXNCRTHQ', 'SMNTRNRTHQ', 'SMNTRARTHQ', 'SMNTERRTHQ', 'SMAMBRTHQ', 'SME1SNSE', 'SME2SNSE', 'SME1ANTI', 'SME2ANTI', 'SME1PCTS', 'SME2PCTS', 'SMGNSDTC', 'SMRRNARD', 'SMRRNART', 'SMMFLGTH', 'SMSFLGTH', 'SMMDFLGTH', 'SMSMFLGTH', 'SMFGCMN', 'SMFGCSD', 'SMFGCSK', 'SMFGCKT', 'SM3PBMN', 'SM3PBSD', 'SM3PBMD', 'SM3PB25P', 'SM3PB75P', 'SM3PBSDM', 'SM3PBGN', 'SMMDMNCV', 'SMMDCVSD', 'SMMDCVCV', 'SMEXCVMD', 'SMEXCVMAD', 'SMMNCV', 'SMUVCRD', 'SMUVCRT', 'SMSHRTRD', 'SMSHRTRT', 'SMSMRDHQ', 'SMSMRTHQ', 'SMPRERDHQ', 'SMPRERTHQ', 'SMSMGNDT', 'SMPREGNDT', 'SMRDLNMN', 'SMRDLNMD', 'SMRDLNSD', 'subject_id', 'sex', 'age', 'DTHHRDY', 'dataset', 'original_sample_id', 'data_type', 'expression_unit', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'original_obs_sample_id', 'tissue_ontology_confidence']\n",
      "\n",
      ".obs.head():\n",
      "                               SMATSSCR SMCENTER SMPTHNTS  rna_integrity_number broad_tissue                                    tissue         SMUBRID  ischemic_time  SMTSPAX array_batch                                 SMNABTCHT SMNABTCHD        batch SMGEBTCHD  SMGEBTCHT   ANALYTE_TYPE SMAFRZE SMGTC      SMRDTTL      SMALTTL     SMALTALG  SMSUPALG  SMRDLGTH     SMVQCFL      SMLMAPQ      SMUMPRD  SMUNPDRD       SMMPPD   SMMAPRT    SMMPPDUN  SMUNMPRT       SMMPDP  SMDPMPRT    SMMPPDXG    SMMPDPXG  SMDPRTXG   SMCHMRD   SMCHMRT    SMMPPDPR    SMMPHQRD  SMMPHQRT    SMMPLQRD  SMSPLTRT    SME1MPRD    SME2MPRD  SME1MPRT  SME2MPRT     SME1MMB     SME2MMB      SME1TTLB  ...  SMMDFLGTH  SMSMFLGTH   SMFGCMN   SMFGCSD   SMFGCSK   SMFGCKT   SM3PBMN   SM3PBSD   SM3PBMD  SM3PB25P  SM3PB75P  SM3PBSDM  SM3PBGN  SMMDMNCV  SMMDCVSD  SMMDCVCV  SMEXCVMD  SMEXCVMAD  SMMNCV  SMUVCRD  SMUVCRT  SMSHRTRD  SMSHRTRT  SMSMRDHQ  SMSMRTHQ  SMPRERDHQ  SMPRERTHQ  SMSMGNDT  SMPREGNDT  SMRDLNMN  SMRDLNMD  SMRDLNSD  \\\n",
      "sample_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ...                                                                                                                                                                                                                                                                                                                                    \n",
      "GTEX-1117F-0005-SM-HL9SH            NaN       B1      nan                   8.3        Blood                               whole blood  UBERON:0013756         1188.0      NaN    BP-38553  RNA isolation PAXgene Blood RNA (Manual)    5/3/13  LCSET-13764       nan  TruSeq.v1  RNA:Total RNA  RNASEQ   nan  134604994.0  367633937.0  232956008.0   72935.0      76.0  16433925.0  304143172.0  127430078.0       0.0  121989620.0  0.957306  14338894.0  0.117542  107650726.0  0.882458  34479795.0  21164891.0  0.613835  133964.0  0.002204  60791324.0  46709230.0  0.382895  75280390.0  0.512743  61034403.0  60955217.0  0.957928  0.956685  13521527.0  11004004.0  4.638615e+09  ...      184.0    77.0952  0.512989  0.088502  0.106281 -0.386421  0.575052  0.311403  0.609091  0.333333  0.862853  0.394237   1532.0  0.089921  0.313438   1.52756  0.542654   0.524121     NaN      NaN      NaN       NaN       NaN       NaN       NaN        NaN        NaN       NaN        NaN       NaN       NaN       NaN   \n",
      "GTEX-1117F-0011-R10b-SM-GI4VE       NaN   B1, A1      nan                   7.2        Brain              brain - frontal cortex (ba9)  UBERON:0009834         1193.0      NaN    BP-42319      RNA Isolation via QIAGEN Spin Column   8/14/13  LCSET-12010       nan  TruSeq.v1  RNA:Total RNA  RNASEQ   nan   67350184.0   75528103.0    8050459.0  127460.0      76.0   3574302.0   12367264.0   64133806.0       0.0   63658224.0  0.992585  41422920.0  0.650708   22235304.0  0.349292  63632092.0  22229041.0  0.349337  345107.0  0.010884  31709272.0  58772285.0  0.923247   4885939.0  0.161648  31884555.0  31773669.0  0.994314  0.990856   5777999.0   6744152.0  2.423226e+09  ...      227.0    85.9908  0.450918  0.088735  0.431654  0.153212  0.677236  0.284687  0.763397  0.500000  0.919192  0.270647   5525.0  1.600000  1.627710   1.00146  0.317303   0.305361     NaN      NaN      NaN       NaN       NaN       NaN       NaN        NaN        NaN       NaN        NaN       NaN       NaN       NaN   \n",
      "GTEX-1117F-0011-R11b-SM-GIN8R       NaN   B1, A1      nan                   6.0        Brain             brain - cerebellar hemisphere  UBERON:0002037         1193.0      NaN    BP-42319      RNA Isolation via QIAGEN Spin Column   8/14/13  LCSET-12011       nan  TruSeq.v1  RNA:Total RNA  RNASEQ   nan   99021352.0  122532321.0   23266593.0  244376.0      76.0   4879614.0   33578345.0   94917928.0       0.0   93852552.0  0.988776  56702303.0  0.604164   37150249.0  0.395836  93805662.0  37136163.0  0.395884  785576.0  0.016839  46651783.0  82287738.0  0.876777  11564814.0  0.150490  47013269.0  46839283.0  0.990609  0.986943   9104653.0  11045926.0  3.573008e+09  ...      220.0    75.6126  0.464407  0.089805  0.479273  0.203122  0.735883  0.258453  0.833333  0.604651  0.936306  0.187664   7779.0  1.685340  1.768670   1.05286  0.326881   0.314845     NaN      NaN      NaN       NaN       NaN       NaN       NaN        NaN        NaN       NaN        NaN       NaN       NaN       NaN   \n",
      "GTEX-1117F-0011-R2b-SM-GI4VL        NaN   B1, A1      nan                   5.7        Brain                  brain - substantia nigra  UBERON:0002038         1193.0      NaN    BP-42208      RNA Isolation via QIAGEN Spin Column   8/12/13  LCSET-12010       nan  TruSeq.v1  RNA:Total RNA  RNASEQ   nan  102541796.0  126416358.0   23734034.0  140528.0      76.0   5688018.0   35903889.0   97733244.0       0.0   96849538.0  0.990958  38788158.0  0.400499   58061380.0  0.599501  96787962.0  58038184.0  0.599643  400397.0  0.008303  48223870.0  83835584.0  0.865627  13013954.0  0.170015  48514920.0  48334618.0  0.992803  0.989113   9121457.0  11355719.0  3.687134e+09  ...      240.0    85.9908  0.453627  0.073098  0.593553  1.329780  0.642670  0.294297  0.720368  0.428571  0.899329  0.312334   4963.0  1.062430  1.370940   1.09517  0.360427   0.348606     NaN      NaN      NaN       NaN       NaN       NaN       NaN        NaN        NaN       NaN        NaN       NaN       NaN       NaN   \n",
      "GTEX-1117F-0011-R3a-SM-GJ3PJ        NaN   B1, A1      nan                   7.1        Brain  brain - anterior cingulate cortex (ba24)  UBERON:0009835         1193.0      NaN    BP-42208      RNA Isolation via QIAGEN Spin Column   8/12/13  LCSET-12012       nan  TruSeq.v1  RNA:Total RNA  RNASEQ   nan   87919278.0  104564329.0   16487319.0  157732.0      76.0   3586415.0   24566291.0   84837998.0       0.0   84118418.0  0.991518  51897218.0  0.616954   32221200.0  0.383046  84088399.0  32214298.0  0.383100  588467.0  0.014062  41848006.0  75020627.0  0.891845   9097791.0  0.174527  42144901.0  41973517.0  0.993538  0.989498   7711933.0   9146972.0  3.203012e+09  ...      226.0    87.4734  0.468294  0.090214  0.518400  0.177329  0.683389  0.286644  0.777778  0.500000  0.920562  0.256071   6600.0  1.650260  1.681710   1.07164  0.328040   0.325567     NaN      NaN      NaN       NaN       NaN       NaN       NaN        NaN        NaN       NaN        NaN       NaN       NaN       NaN   \n",
      "\n",
      "                               subject_id      sex    age  DTHHRDY  dataset  original_sample_id  data_type  expression_unit  species  species_ontology                           tissue_original  tissue_ontology  data_type_original  assay_ontology  age_original  developmental_stage_ontology         original_obs_sample_id  tissue_ontology_confidence  \n",
      "sample_id                                                                                                                                                                                                                                                                                                                                                     \n",
      "GTEX-1117F-0005-SM-HL9SH       GTEX-1117F  unknown  60-69      4.0     GTEx                        RNA-seq                     human    NCBITaxon:9606                               Whole Blood   UBERON:0000178                         EFO:0009922         60-69                                     GTEX-1117F-0005-SM-HL9SH                      medium  \n",
      "GTEX-1117F-0011-R10b-SM-GI4VE  GTEX-1117F  unknown  60-69      4.0     GTEx                        RNA-seq                     human    NCBITaxon:9606              Brain - Frontal Cortex (BA9)   UBERON:0013529                         EFO:0009922         60-69                                GTEX-1117F-0011-R10b-SM-GI4VE                      medium  \n",
      "GTEX-1117F-0011-R11b-SM-GIN8R  GTEX-1117F  unknown  60-69      4.0     GTEx                        RNA-seq                     human    NCBITaxon:9606             Brain - Cerebellar Hemisphere   UBERON:0002245                         EFO:0009922         60-69                                GTEX-1117F-0011-R11b-SM-GIN8R                      medium  \n",
      "GTEX-1117F-0011-R2b-SM-GI4VL   GTEX-1117F  unknown  60-69      4.0     GTEx                        RNA-seq                     human    NCBITaxon:9606                  Brain - Substantia nigra   UBERON:0002038                         EFO:0009922         60-69                                 GTEX-1117F-0011-R2b-SM-GI4VL                      medium  \n",
      "GTEX-1117F-0011-R3a-SM-GJ3PJ   GTEX-1117F  unknown  60-69      4.0     GTEx                        RNA-seq                     human    NCBITaxon:9606  Brain - Anterior cingulate cortex (BA24)   UBERON:0009835                         EFO:0009922         60-69                                 GTEX-1117F-0011-R3a-SM-GJ3PJ                      medium  \n",
      "\n",
      "[5 rows x 136 columns]\n",
      "\n",
      "Unique values in key .obs columns:\n",
      "  'dataset': ['GTEx'] (1 total unique)\n",
      "  'tissue': ['whole blood', 'brain - frontal cortex (ba9)', 'brain - cerebellar hemisphere', 'brain - substantia nigra', 'brain - anterior cingulate cortex (ba24)', 'brain - amygdala', 'brain - caudate (basal ganglia)', 'brain - nucleus accumbens (basal ganglia)', 'brain - putamen (basal ganglia)', 'adipose - subcutaneous']... (54 total unique)\n",
      "  'sex': ['unknown'] (1 total unique)\n",
      "  'data_type': ['RNA-seq'] (1 total unique)\n",
      "  'expression_unit': [''] (1 total unique)\n",
      "  'tissue_ontology': ['UBERON:0000178', 'UBERON:0013529', 'UBERON:0002245', 'UBERON:0002038', 'UBERON:0009835', 'UBERON:0001876', 'UBERON:0001873', 'UBERON:0001882', 'UBERON:0001874', 'UBERON:0002190']... (53 total unique)\n",
      "\n",
      "Missing values count in key .obs columns:\n",
      "{'tissue_ontology': np.int64(979)}\n",
      "------------------------------\n",
      "\n",
      "--- Variables (.var) Info (GTEX) ---\n",
      ".var shape: (58988, 9)\n",
      ".var columns: ['gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence']\n",
      ".var index name: None\n",
      "First 5 var_names (index): ['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460']\n",
      "\n",
      ".var.head():\n",
      "                         gene_id        original_ids gene_name       gene_type chromosome     mapping_source original_gene_id       ensembl_id mapping_confidence\n",
      "ENSG00000000003  ENSG00000000003  ENSG00000000003.15    TSPAN6  protein_coding       chrX  reference_mapping  ENSG00000000003  ENSG00000000003               high\n",
      "ENSG00000000005  ENSG00000000005   ENSG00000000005.6      TNMD  protein_coding       chrX  reference_mapping  ENSG00000000005  ENSG00000000005               high\n",
      "ENSG00000000419  ENSG00000000419  ENSG00000000419.14      DPM1  protein_coding      chr20  reference_mapping  ENSG00000000419  ENSG00000000419               high\n",
      "ENSG00000000457  ENSG00000000457  ENSG00000000457.14     SCYL3  protein_coding       chr1  reference_mapping  ENSG00000000457  ENSG00000000457               high\n",
      "ENSG00000000460  ENSG00000000460  ENSG00000000460.17  C1orf112  protein_coding       chr1  reference_mapping  ENSG00000000460  ENSG00000000460               high\n",
      "\n",
      "Unique values in key .var columns:\n",
      "  'gene_type': ['protein_coding', 'unitary_pseudogene', 'unprocessed_pseudogene', 'processed_pseudogene', 'transcribed_unprocessed_pseudogene', 'processed_transcript', 'antisense', 'polymorphic_pseudogene', 'lincRNA', 'sense_intronic']... (45 total unique)\n",
      "  'chromosome': ['chrX', 'chr20', 'chr1', 'chr6', 'chr3', 'chr7', 'chr12', 'chr11', 'chr4', 'chr17']... (26 total unique)\n",
      "  'mapping_source': ['reference_mapping', 'unmapped'] (2 total unique)\n",
      "  'mapping_confidence': ['high', 'none'] (2 total unique)\n",
      "------------------------------\n",
      "\n",
      "--- Unstructured Annotation (.uns) Info (GTEX) ---\n",
      ".uns keys: ['dataset_info', 'dataset_version', 'gencode_version', 'gene_mapping_stats', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'original_gencode_version', 'original_reference_genome', 'processing_date', 'reference_genome', 'rna_seq_protocol']\n",
      "\n",
      ".uns['dataset_info']: \n",
      "{\n",
      "  \"data_type\": \"RNA-seq\",\n",
      "  \"expression_unit\": \"TPM\",\n",
      "  \"gencode_version\": \"24\",\n",
      "  \"genes\": \"58988\",\n",
      "  \"samples\": \"19616\",\n",
      "  \"source\": \"GTEx\",\n",
      "  \"subject_count\": \"946\",\n",
      "  \"tissue_count\": \"54\",\n",
      "  \"version\": \"v10\"\n",
      "}\n",
      "\n",
      ".uns['gene_mapping_stats']: \n",
      "{\n",
      "  \"mapped_genes\": \"54754\",\n",
      "  \"mapping_percentage\": 92.82226893605478,\n",
      "  \"mapping_source\": \"reference_mapping\",\n",
      "  \"total_genes\": \"58988\"\n",
      "}\n",
      "\n",
      ".uns['metadata_validation']: \n",
      "{\n",
      "  \"dataset\": \"GTEx\",\n",
      "  \"missing_fields\": \"[]\",\n",
      "  \"missing_values\": {\n",
      "    \"DTHHRDY\": \"113\",\n",
      "    \"SMATSSCR\": \"4536\",\n",
      "    \"SMGEBTCHD\": \"2981\",\n",
      "    \"SMGTC\": \"19616\",\n",
      "    \"SMMNCV\": \"19616\",\n",
      "    \"SMPREGNDT\": \"19616\",\n",
      "    \"SMPRERDHQ\": \"19616\",\n",
      "    \"SMPRERTHQ\": \"19616\",\n",
      "    \"SMPTHNTS\": \"4537\",\n",
      "    \"SMRDLNMD\": \"19616\",\n",
      "    \"SMRDLNMN\": \"19616\",\n",
      "    \"SMRDLNSD\": \"19616\",\n",
      "    \"SMSHRTRD\": \"19616\",\n",
      "    \"SMSHRTRT\": \"19616\",\n",
      "    \"SMSMGNDT\": \"19616\",\n",
      "    \"SMSMRDHQ\": \"19616\",\n",
      "    \"SMSMRTHQ\": \"19616\",\n",
      "    \"SMTSPAX\": \"4557\",\n",
      "    \"SMUVCRD\": \"19616\",\n",
      "    \"SMUVCRT\": \"19616\",\n",
      "    \"batch\": \"11\",\n",
      "    \"ischemic_time\": \"24\"\n",
      "  },\n",
      "  \"sample_count\": \"19616\",\n",
      "  \"unmapped_ages\": \"['60-69' '50-59' '40-49' '20-29' '30-39' '70-79']\",\n",
      "  \"unmapped_assays\": \"[]\",\n",
      "  \"unmapped_tissues\": \"[]\",\n",
      "  \"validation_status\": \"warning\"\n",
      "}\n",
      "\n",
      ".uns['harmonized_gencode_version']: \n",
      "24\n",
      "\n",
      ".uns['harmonized_reference_genome']: \n",
      "hg38\n",
      "------------------------------\n",
      "\n",
      "\n",
      "========================= Inspecting: MAGE =========================\n",
      "\n",
      "--- AnnData Object Info (MAGE) ---\n",
      "AnnData object with n_obs × n_vars = 731 × 19428\n",
      "    obs: 'sample_id', 'donor_id', 'subject_id', 'tissue', 'dataset', 'data_type', 'expression_unit', 'is_gencode', 'is_ensembl', 'sex', 'age', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence'\n",
      "    var: 'gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence'\n",
      "    uns: 'assay_ontology', 'cell_type_info', 'data_type', 'dataset_info', 'gencode_version', 'gene_mapping_stats', 'genome_info', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'ontology_mappings', 'platform', 'processing_date', 'reference_genome', 'rna_seq_protocol'\n",
      "------------------------------\n",
      "\n",
      "--- Observations (.obs) Info (MAGE) ---\n",
      ".obs shape: (731, 20)\n",
      ".obs columns: ['sample_id', 'donor_id', 'subject_id', 'tissue', 'dataset', 'data_type', 'expression_unit', 'is_gencode', 'is_ensembl', 'sex', 'age', 'species', 'species_ontology', 'tissue_original', 'tissue_ontology', 'data_type_original', 'assay_ontology', 'age_original', 'developmental_stage_ontology', 'tissue_ontology_confidence']\n",
      "\n",
      ".obs.head():\n",
      "                       sample_id donor_id subject_id       tissue dataset data_type expression_unit is_gencode is_ensembl      sex age species species_ontology tissue_original tissue_ontology data_type_original assay_ontology age_original developmental_stage_ontology tissue_ontology_confidence\n",
      "_sample_id                                                                                                                                                                                                                                                                                            \n",
      "NA06985_NA06985  NA06985_NA06985  NA06985    NA06985  lymphoblast    MAGE   rna-seq             TPM      False       True  unknown       human   NCBITaxon:9606     lymphoblast                            RNA-seq    EFO:0009922                                                                 none\n",
      "NA07000_NA07000  NA07000_NA07000  NA07000    NA07000  lymphoblast    MAGE   rna-seq             TPM      False       True  unknown       human   NCBITaxon:9606     lymphoblast                            RNA-seq    EFO:0009922                                                                 none\n",
      "NA11919_NA11919  NA11919_NA11919  NA11919    NA11919  lymphoblast    MAGE   rna-seq             TPM      False       True  unknown       human   NCBITaxon:9606     lymphoblast                            RNA-seq    EFO:0009922                                                                 none\n",
      "NA11930_NA11930  NA11930_NA11930  NA11930    NA11930  lymphoblast    MAGE   rna-seq             TPM      False       True  unknown       human   NCBITaxon:9606     lymphoblast                            RNA-seq    EFO:0009922                                                                 none\n",
      "NA11932_NA11932  NA11932_NA11932  NA11932    NA11932  lymphoblast    MAGE   rna-seq             TPM      False       True  unknown       human   NCBITaxon:9606     lymphoblast                            RNA-seq    EFO:0009922                                                                 none\n",
      "\n",
      "Unique values in key .obs columns:\n",
      "  'dataset': ['MAGE'] (1 total unique)\n",
      "  'tissue': ['lymphoblast'] (1 total unique)\n",
      "  'sex': ['unknown'] (1 total unique)\n",
      "  'data_type': ['rna-seq'] (1 total unique)\n",
      "  'expression_unit': ['TPM'] (1 total unique)\n",
      "  'tissue_ontology': [''] (1 total unique)\n",
      "\n",
      "Missing values count in key .obs columns:\n",
      "{'age': np.int64(731), 'tissue_ontology': np.int64(731)}\n",
      "------------------------------\n",
      "\n",
      "--- Variables (.var) Info (MAGE) ---\n",
      ".var shape: (19428, 9)\n",
      ".var columns: ['gene_id', 'original_ids', 'gene_name', 'gene_type', 'chromosome', 'mapping_source', 'original_gene_id', 'ensembl_id', 'mapping_confidence']\n",
      ".var index name: None\n",
      "First 5 var_names (index): ['ENSG00000000003', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460', 'ENSG00000000938']\n",
      "\n",
      ".var.head():\n",
      "                         gene_id        original_ids gene_name       gene_type chromosome     mapping_source original_gene_id       ensembl_id mapping_confidence\n",
      "ENSG00000000003  ENSG00000000003  ENSG00000000003.14    TSPAN6  protein_coding       chrX  reference_mapping  ENSG00000000003  ENSG00000000003               high\n",
      "ENSG00000000419  ENSG00000000419  ENSG00000000419.12      DPM1  protein_coding      chr20  reference_mapping  ENSG00000000419  ENSG00000000419               high\n",
      "ENSG00000000457  ENSG00000000457  ENSG00000000457.13     SCYL3  protein_coding       chr1  reference_mapping  ENSG00000000457  ENSG00000000457               high\n",
      "ENSG00000000460  ENSG00000000460  ENSG00000000460.16  C1orf112  protein_coding       chr1  reference_mapping  ENSG00000000460  ENSG00000000460               high\n",
      "ENSG00000000938  ENSG00000000938  ENSG00000000938.12       FGR  protein_coding       chr1  reference_mapping  ENSG00000000938  ENSG00000000938               high\n",
      "\n",
      "Unique values in key .var columns:\n",
      "  'gene_type': ['protein_coding', 'transcribed_unprocessed_pseudogene', 'unprocessed_pseudogene', 'unitary_pseudogene', 'processed_transcript', 'lincRNA', 'processed_pseudogene', 'sense_intronic', 'antisense', 'transcribed_processed_pseudogene']... (32 total unique)\n",
      "  'chromosome': ['chrX', 'chr20', 'chr1', 'chr6', 'chr3', 'chr7', 'chr12', 'chr11', 'chr4', 'chr17']... (23 total unique)\n",
      "  'mapping_source': ['reference_mapping'] (1 total unique)\n",
      "  'mapping_confidence': ['high'] (1 total unique)\n",
      "------------------------------\n",
      "\n",
      "--- Unstructured Annotation (.uns) Info (MAGE) ---\n",
      ".uns keys: ['assay_ontology', 'cell_type_info', 'data_type', 'dataset_info', 'gencode_version', 'gene_mapping_stats', 'genome_info', 'harmonized_gencode_version', 'harmonized_reference_genome', 'metadata_sources', 'metadata_validation', 'ontology_mappings', 'platform', 'processing_date', 'reference_genome', 'rna_seq_protocol']\n",
      "\n",
      ".uns['dataset_info']: \n",
      "{\n",
      "  \"accession\": \"PRJNA851328\",\n",
      "  \"cell_type\": \"lymphoblastoid cell lines\",\n",
      "  \"continental_groups\": \"5\",\n",
      "  \"data_repository\": \"[{\\\"name\\\": \\\"Sequence Read Archive (SRA)\\\", \\\"accession\\\": \\\"PRJNA851328\\\"}, {\\\"name\\\": \\\"Zenodo\\\", \\\"doi\\\": \\\"10.5281/zenodo.10535719\\\"}]\",\n",
      "  \"library_count\": \"779\",\n",
      "  \"populations\": \"26\",\n",
      "  \"sample_size\": \"731\"\n",
      "}\n",
      "\n",
      ".uns['gene_mapping_stats']: \n",
      "{\n",
      "  \"mapped_genes\": \"19428\",\n",
      "  \"mapping_percentage\": 100.0,\n",
      "  \"mapping_source\": \"reference_mapping\",\n",
      "  \"total_genes\": \"19428\"\n",
      "}\n",
      "\n",
      ".uns['metadata_validation']: \n",
      "{\n",
      "  \"dataset\": \"MAGE\",\n",
      "  \"missing_fields\": \"[]\",\n",
      "  \"missing_values\": {},\n",
      "  \"sample_count\": \"731\",\n",
      "  \"unmapped_ages\": \"[]\",\n",
      "  \"unmapped_assays\": \"[]\",\n",
      "  \"unmapped_tissues\": \"[]\",\n",
      "  \"validation_status\": \"warning\"\n",
      "}\n",
      "\n",
      ".uns['harmonized_gencode_version']: \n",
      "24\n",
      "\n",
      ".uns['harmonized_reference_genome']: \n",
      "hg38\n",
      "------------------------------\n",
      "\n",
      "========================= Inspection Complete =========================\n"
     ]
    }
   ],
   "source": [
    "# %% [markdown]\n",
    "# # Load and Inspect Preprocessed RNA-seq Datasets\n",
    "#\n",
    "# This notebook loads the AnnData files generated by Stage 2.5 of the pipeline\n",
    "# (`preprocess_dataset_gene_ids.py`) and performs basic checks and inspections.\n",
    "\n",
    "# %%\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np # For isnan check\n",
    "\n",
    "# Configure pandas display options for better viewing\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.width', 1000)\n",
    "\n",
    "# %%\n",
    "# --- Configuration ---\n",
    "# Directory containing the preprocessed h5ad files\n",
    "preprocessed_dir = \"/mnt/czi-sci-ai/intrinsic-variation-gene-ex/rnaseq/preprocessed_data/run_20250502_211754/\"\n",
    "\n",
    "# List of dataset prefixes to load\n",
    "dataset_names = ['adni', 'encode', 'gtex', 'mage'] # Add 'entex' if it was processed\n",
    "\n",
    "# --- End Configuration ---\n",
    "\n",
    "# %%\n",
    "# Dictionary to hold the loaded AnnData objects\n",
    "adata_dict = {}\n",
    "\n",
    "print(f\"Loading datasets from: {preprocessed_dir}\\n\")\n",
    "\n",
    "for name in dataset_names:\n",
    "    file_name = f\"{name}_standardized_preprocessed.h5ad\"\n",
    "    file_path = os.path.join(preprocessed_dir, file_name)\n",
    "    print(f\"--- Loading {name.upper()} ---\")\n",
    "    print(f\"File: {file_path}\")\n",
    "    if os.path.exists(file_path):\n",
    "        try:\n",
    "            adata_dict[name] = sc.read_h5ad(file_path)\n",
    "            print(f\"Successfully loaded {name.upper()} with shape {adata_dict[name].shape}\")\n",
    "        except Exception as e:\n",
    "            print(f\"ERROR: Could not load {file_path}: {e}\")\n",
    "    else:\n",
    "        print(f\"WARNING: File not found: {file_path}\")\n",
    "    print(\"-\" * (len(name) + 14))\n",
    "\n",
    "# %% [markdown]\n",
    "# ## Inspect Loaded Datasets\n",
    "\n",
    "# %%\n",
    "if not adata_dict:\n",
    "    print(\"No datasets were loaded successfully. Cannot perform inspection.\")\n",
    "else:\n",
    "    for name, adata in adata_dict.items():\n",
    "        print(f\"\\n\\n========================= Inspecting: {name.upper()} =========================\\n\")\n",
    "\n",
    "        # --- Basic Info ---\n",
    "        print(f\"--- AnnData Object Info ({name.upper()}) ---\")\n",
    "        print(adata)\n",
    "        print(\"-\" * 30)\n",
    "\n",
    "        # --- Observations (.obs) ---\n",
    "        print(f\"\\n--- Observations (.obs) Info ({name.upper()}) ---\")\n",
    "        print(f\".obs shape: {adata.obs.shape}\")\n",
    "        print(f\".obs columns: {adata.obs.columns.tolist()}\")\n",
    "        print(\"\\n.obs.head():\")\n",
    "        print(adata.obs.head())\n",
    "        # Check unique values for key categorical fields\n",
    "        print(\"\\nUnique values in key .obs columns:\")\n",
    "        for col in ['dataset', 'tissue', 'sex', 'data_type', 'expression_unit', 'tissue_ontology']:\n",
    "            if col in adata.obs.columns:\n",
    "                try:\n",
    "                    unique_vals = adata.obs[col].unique()\n",
    "                    # Handle categorical output nicely\n",
    "                    if isinstance(unique_vals, pd.Categorical):\n",
    "                         unique_vals = unique_vals.tolist()\n",
    "                    # Limit display if too many unique values (e.g., tissue in GTEx)\n",
    "                    display_vals = unique_vals[:10] if len(unique_vals) > 10 else unique_vals\n",
    "                    ellipsis = \"...\" if len(unique_vals) > 10 else \"\"\n",
    "                    print(f\"  '{col}': {display_vals}{ellipsis} ({len(unique_vals)} total unique)\")\n",
    "                except Exception as e:\n",
    "                    print(f\"  Could not get unique values for '{col}': {e}\")\n",
    "            else:\n",
    "                print(f\"  '{col}': Column not found\")\n",
    "        # Check for missing values\n",
    "        print(\"\\nMissing values count in key .obs columns:\")\n",
    "        missing_data = {}\n",
    "        for col in ['tissue', 'subject_id', 'dataset', 'sex', 'age', 'tissue_ontology']:\n",
    "             if col in adata.obs.columns:\n",
    "                # Count NaNs directly\n",
    "                na_count = adata.obs[col].isna().sum()\n",
    "                # Count empty strings '', being careful about NaNs\n",
    "                is_empty_series = adata.obs[col].apply(lambda x: isinstance(x, str) and x.strip() == '')\n",
    "                try:\n",
    "                    empty_count = np.sum(is_empty_series.to_numpy(dtype=bool))\n",
    "                except: # Fallback if conversion fails\n",
    "                    empty_count = sum(1 for x in is_empty_series if x is True)\n",
    "                total_missing = na_count + empty_count\n",
    "                if total_missing > 0:\n",
    "                    missing_data[col] = total_missing\n",
    "        if missing_data:\n",
    "            print(missing_data)\n",
    "        else:\n",
    "             print(\"  No missing/empty values found in checked key columns.\")\n",
    "        print(\"-\" * 30)\n",
    "\n",
    "\n",
    "        # --- Variables (.var) ---\n",
    "        print(f\"\\n--- Variables (.var) Info ({name.upper()}) ---\")\n",
    "        print(f\".var shape: {adata.var.shape}\")\n",
    "        print(f\".var columns: {adata.var.columns.tolist()}\")\n",
    "        print(f\".var index name: {adata.var.index.name}\")\n",
    "        print(f\"First 5 var_names (index): {adata.var_names[:5].tolist()}\")\n",
    "        print(\"\\n.var.head():\")\n",
    "        print(adata.var.head())\n",
    "         # Check unique values for key categorical fields\n",
    "        print(\"\\nUnique values in key .var columns:\")\n",
    "        for col in ['gene_type', 'chromosome', 'mapping_source', 'mapping_confidence']:\n",
    "            if col in adata.var.columns:\n",
    "                try:\n",
    "                    unique_vals = adata.var[col].unique()\n",
    "                     # Handle categorical output nicely\n",
    "                    if isinstance(unique_vals, pd.Categorical):\n",
    "                         unique_vals = unique_vals.tolist()\n",
    "                    display_vals = unique_vals[:10] if len(unique_vals) > 10 else unique_vals\n",
    "                    ellipsis = \"...\" if len(unique_vals) > 10 else \"\"\n",
    "                    print(f\"  '{col}': {display_vals}{ellipsis} ({len(unique_vals)} total unique)\")\n",
    "                except Exception as e:\n",
    "                    print(f\"  Could not get unique values for '{col}': {e}\")\n",
    "            else:\n",
    "                 print(f\"  '{col}': Column not found\")\n",
    "        print(\"-\" * 30)\n",
    "\n",
    "        # --- Unstructured Annotation (.uns) ---\n",
    "        print(f\"\\n--- Unstructured Annotation (.uns) Info ({name.upper()}) ---\")\n",
    "        print(f\".uns keys: {list(adata.uns.keys())}\")\n",
    "        # Print specific .uns fields if they exist\n",
    "        for key in ['dataset_info', 'gene_mapping_stats', 'metadata_validation', 'harmonized_gencode_version', 'harmonized_reference_genome']:\n",
    "            if key in adata.uns:\n",
    "                print(f\"\\n.uns['{key}']: \")\n",
    "                # Pretty print dictionaries\n",
    "                if isinstance(adata.uns[key], dict):\n",
    "                    import json\n",
    "                    print(json.dumps(adata.uns[key], indent=2, default=str)) # Use default=str for non-serializable\n",
    "                else:\n",
    "                    print(adata.uns[key])\n",
    "        print(\"-\" * 30)\n",
    "\n",
    "print(\"\\n========================= Inspection Complete =========================\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
